impute_constant/impute_constant.tex

```diff
@@ -0,0 +1,181 @@
+        \item \texttt{DA::NonNull} must have traits \texttt{Clone}
+    \end{itemize}
+\end{itemize}
+
+\subsubsection*{Postconditions}
+
+\begin{itemize}
+    \item Either a valid \texttt{Transformation} is returned or an error is returned.
+\end{itemize}
+
+
+\begin{lstlisting}[language=Python,  escapechar=|]
+def make_impute_constant(constant : DA::NonNull):
+    # instead of VectorDomain(DA), we add ::new() to get a new instance of DA. This is bcause DA has the ImputableDomain trait. Discuss among interns.
+    input_domain = VectorDomain(DA::new())
+    output_domain = VectorDomain(AllDomain(DA::NonNull))
+    input_metric = SymmetricDistance()
+    output_metric = SymmetricDistance()
+    
+    # check constant for nullity first
+    assert(not constant.is_null); # not DA::is_null(constant) |\label{line:null}|
+    def Relation(d_in: u32, d_out: u32) -> bool: |\label{line:rel}|
+        return d_out >= d_in*1
+
+    def function(data: Vec[DA::Carrier]) -> Vec[DA::NonNull]: |\label{line:fn}|
+        def impute_constant(x: DA) -> DA::NonNull: |\label{line:atom-fn}|
+            return constant if x.is_null else x
+        return list(map(impute_constant, data)) |\label{line:map}|
+        
+    return Transformation(input_domain, output_domain, function, input_metric, output_metric, stability_relation=Relation)
+    # return make_row_by_row(input_domain, output_domain, impute_constant);
+\end{lstlisting}
+
+\grace{ Will need to change pseudocode so that it returns the result of a make row by row transformation (which the code does) instead of a Transformation directly.}
+
+\subsection{Proof}
+\grace{Should this go in the proofs list?}
+\grace{I'm also confused why we have implementation of DA::NonNull but it doesn't do what the name says. Seems pretty counter intuitive.}
+
+
+\begin{lemma}[\texttt{DA::NonNull} contains \texttt{null}] \label{lemma-NonNull}
+\texttt{var} of type \texttt{DA::NonNull} can be of type \texttt{null}.
+\end{lemma}
+\begin{proof}
+Let the domain of atom variable \texttt{DA} be \texttt{InherentNullDomain<AllDomain<f64>>}. Recall that \texttt{InherentNullDomain} exists for types that can represent null inherently in the carrier type. Then the type $$\texttt{DA::NonNull == InherentNullDomain<AllDomain<f64>>::NonNull == f64}.$$ The latter holds because in the \texttt{InherentNullDomain} implementation in the rust code \url{https://github.com/opendp/opendp/blob/main/rust/opendp/src/trans/impute.rs#L48-L56}, the \texttt{type NonNull = Self::Carrier}. The \texttt{Carrier} of \texttt{VectorDomain<AllDomain<T>>} has type \texttt{T}, so in this case the \texttt{::Carrier} is type \texttt{f64}.
+
+Therefore \texttt{var} is also of type \texttt{f64}. \texttt{f64} can contain \texttt{null} values, so we are done.
+\end{proof}
```

Comment:
3. is it correct that a variable with type DA::NonNull can contain null values?

Reply:
The type may be capable of representing a null value. If an instance of DA::NonNull is null, it is either imputed or an construction-time error is raised by 2.